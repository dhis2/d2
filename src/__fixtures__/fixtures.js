export default (function x() {
    const fixtures = {};

    function getFixture(fixtureName) {
        if (fixtures && fixtures[fixtureName]) {
            // Make sure we return a new object
            return JSON.parse(JSON.stringify(fixtures[fixtureName]));
        }
        throw new Error(['Fixture', fixtureName, 'does not exist'].join(' '));
    }

    function addFixture(fixtureName, fixtureData) {
        if (fixtureName && fixtureData && typeof fixtureName === 'string') {
            fixtures[fixtureName] = fixtureData;
        }
    }

    addFixture('/api/attributes', require('./json/api/schemas/attributes.json'));
    addFixture('/api/dataElements', require('./json/api/dataElements.json'));
    addFixture('/api/organisationUnits', require('./json/api/organisationUnits.json'));
    addFixture('/api/schemas', require('./json/api/schemas.json'));
    addFixture('/api/schemas/dataElement', require('./json/api/schemas/dataElement.json'));
    addFixture('/api/schemas/dataSet', require('./json/api/schemas/dataSet'));
    addFixture('/api/schemas/indicatorGroup', require('./json/api/schemas/indicatorGroup.json'));
    addFixture('/api/schemas/organisationUnit', require('./json/api/schemas/organisationUnit.json'));
    addFixture('/api/schemas/organisationUnitGroupSet', require('./json/api/schemas/organisationUnitGroupSet'));
    addFixture('/api/schemas/validationRule', require('./json/api/schemas/validationRule'));
    addFixture('/api/schemas/program', require('./json/api/schemas/program'));
    addFixture('/api/schemas/programNotificationTemplate', require('./json/api/schemas/programNotificationTemplate'));
    addFixture('/api/schemas/user', require('./json/api/schemas/user.json'));
    addFixture('/api/schemas/legendSet', require('./json/api/schemas/legendSet.json'));
    addFixture('/api/schemas/legend', require('./json/api/schemas/legend.json'));
    addFixture('/api/userSettings', require('./json/api/userSettings'));
    addFixture('/appStore', require('./json/api/appStore.json'));
    addFixture('/dataElementAttributes', require('./json/dataElementAttributes.json'));
    addFixture('/me', require('./json/api/me.json'));
    addFixture('/me/authorities', require('./json/api/me/authorities.json'));
    addFixture('/modelDefinitions/dataElement', require('./json/modelDefinitions/dataElement.json'));
    addFixture('/singleUserAllFields', require('./json/singleUserAllFields.json'));
    addFixture('/singleUserOwnerFields', require('./json/singleUserOwnerFields.json'));
    addFixture('/api/legendSets/k1JHPfXsJND', require('./json/api/legendSets/k1JHPfXsJND.json'));
    addFixture('/api/analytics/dataValueSet', require('./json/api/analytics/dataValueSet.json'));
    addFixture('/api/analytics/rawData', require('./json/api/analytics/rawData.json'));
    addFixture('/api/analytics/aggregate', require('./json/api/analytics/aggregate.json'));
    addFixture('/api/analytics/count', require('./json/api/analytics/count.json'));
    addFixture('/api/analytics/query', require('./json/api/analytics/query.json'));
    addFixture('/api/analytics/cluster', require('./json/api/analytics/cluster.json'));

    return {
        get: getFixture,
        add: addFixture,
    };
}());
