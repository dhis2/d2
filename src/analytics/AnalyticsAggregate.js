import AnalyticsBase from './AnalyticsBase';

/**
 * @extends module:analytics.AnalyticsBase
 *
 * @description
 * Analytics aggregate class used to request aggregate analytics data from Web API.
 *
 * @memberof module:analytics
 *
 * @see https://docs.dhis2.org/master/en/developer/html/webapi_analytics.html
 */
class AnalyticsAggregate extends AnalyticsBase {
    /**
     * @param {!AnalyticsRequest} req Request object
     *
     * @returns {Promise} Promise that resolves with the analytics data value set data from the API.
     *
     * @example
     * const req = new d2.analytics.request()
     *  .addDataDimension(['Uvn6LCg7dVU','OdiHJayrsKo'])
     *  .addPeriodDimension('LAST_4_QUARTERS')
     *  .addOrgUnitDimension(['lc3eMKXaEfw','PMa2VCrupOd'])
     *  .addOrgUnitFilter('O6uvpzGd5pu')
     *  .withStartDate('2017-10-01')
     *  .withEndDate('2017-10-31');
     *
     * d2.analytics.aggregate.getDataValueSet(req)
     * .then(console.log);
     */
    getDataValueSet(req) {
        return this.get(req.withPath('dataValueSet'));
    }

    /**
     * @param {!AnalyticsRequest} req Request object
     *
     * @returns {Promise} Promise that resolves with the raw data from the API.
     *
     * @example
     * const req = new d2.analytics.request()
     *  .addDataDimension(['Uvn6LCg7dVU','OdiHJayrsKo'])
     *  .addPeriodDimension('LAST_4_QUARTERS')
     *  .addOrgUnitDimension(['lc3eMKXaEfw', 'PMa2VCrupOd'])
     *  .addOrgUnitFilter('O6uvpzGd5pu');
     *  .withStartDate('2017-10-01')
     *  .withEndDate('2017-10-31')
     *  .withFormat('xml');
     *
     *  d2.analytics.aggregate.getRawData(req)
     *  .then(console.log);
     */
    getRawData(req) {
        return this.get(req.withPath('rawData'));
    }

    /**
     * @param {!AnalyticsRequest} req Request object
     *
     * @returns {Promise} Promise that resolves with the SQL statement used to query the database.
     *
     * @example
     * const req = new d2.analytics.request()
     *  .addDataDimension(['Uvn6LCg7dVU','OdiHJayrsKo'])
     *  .addPeriodDimension('LAST_4_QUARTERS')
     *  .addOrgUnitDimension(['lc3eMKXaEfw', 'PMa2VCrupOd'])
     *  .addOrgUnitFilter('O6uvpzGd5pu');
     *  .withStartDate('2017-10-01')
     *  .withEndDate('2017-10-31');
     *
     *  d2.analytics.aggregate.getDebugSql(req);
     *  .then(console.log);
     */
    getDebugSql(req) {
        return this.get(req.withPath('debug/sql'));
    }
}

export default AnalyticsAggregate;
