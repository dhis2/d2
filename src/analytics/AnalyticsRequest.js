import AnalyticsRequestDimensionsMixin from './AnalyticsRequestDimensionsMixin';
import AnalyticsRequestFiltersMixin from './AnalyticsRequestFiltersMixin';
import AnalyticsRequestPropertiesMixin from './AnalyticsRequestPropertiesMixin';
import AnalyticsRequestBase from './AnalyticsRequestBase';

/**
 * @description
 * Class for constructing a request object to use for communicating with the analytics API endpoint.
 *
 * @param {!Object} options Object with analytics request options
 *
 * @memberof module:analytics
 *
 * @extends module:analytics.AnalyticsRequestDimensionsMixin
 * @extends module:analytics.AnalyticsRequestFiltersMixin
 * @extends module:analytics.AnalyticsRequestPropertiesMixin
 * @extends module:analytics.AnalyticsRequestBase
 */
class AnalyticsRequest extends AnalyticsRequestDimensionsMixin(
    AnalyticsRequestFiltersMixin(AnalyticsRequestPropertiesMixin(AnalyticsRequestBase)),
) {
    /**
     * Extracts dimensions and filters from an analytic object model and add them to the request
     *
     * @param {Object} model The analytics object model from which extract the dimensions/filters
     *
     * @returns {AnalyticsRequest} A new instance of the class for chaining purposes
     *
     * @example
     * const req = new d2.analytics.request()
     *    .fromModel(model);
     *
     * // dimension=pe:LAST_12_MONTH&dimension=dx:fbfJHSPpUQD;cYeuwXTCPkU;Jtf34kNZhzP;hfdmMSPBgLG&filter=ou:ImspTQPwCqd
     */
    fromModel(model) {
        let request = this;

        // extract dimensions from model
        const columns = model.columns || [];
        const rows = model.rows || [];

        columns.concat(rows).forEach((d) => {
            let dimension = d.dimension;

            if (d.filter) {
                dimension += `:${d.filter}`;
            }

            request = request.addDimension(dimension, d.items.map(item => item.id));
        });

        // extract filters from model
        const filters = model.filters || [];

        filters.forEach(
            f => (request = request.addFilter(f.dimension, f.items.map(item => item.id))),
        );

        return request;
    }
}

export default AnalyticsRequest;
