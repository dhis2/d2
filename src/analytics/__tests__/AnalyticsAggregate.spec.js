import fixtures from '../../__fixtures__/fixtures';
import MockApi from '../../api/Api';
import AnalyticsRequest from '../AnalyticsRequest';
import AnalyticsAggregate from '../AnalyticsAggregate';

jest.mock('../../api/Api'); // src/api/__mocks/Api.js

const debugSqlFixture = `select de.name as de_name, de.uid as de_uid, de.dataelementid as de_id, pe.startdate as
start_date, pe.enddate as end_date, pt.name as pt_name, ou.name as ou_name, ou.uid as ou_uid, ou.organisationunitid as
ou_id, coc.name as coc_name, coc.uid as coc_uid, coc.categoryoptioncomboid as coc_id, aoc.name as aoc_name, aoc.uid as
aoc_uid, aoc.categoryoptioncomboid as aoc_id, dv.value as datavalue from datavalue dv inner join dataelement de on
dv.dataelementid = de.dataelementid inner join period pe on dv.periodid = pe.periodid inner join periodtype pt on
pe.periodtypeid = pt.periodtypeid inner join organisationunit ou on dv.sourceid = ou.organisationunitid inner join
categoryoptioncombo coc on dv.categoryoptioncomboid = coc.categoryoptioncomboid inner join categoryoptioncombo aoc on
dv.attributeoptioncomboid = aoc.categoryoptioncomboid where dv.dataelementid in (359596,359597) and ((pe.startdate >=
    '2016-01-01' and pe.enddate <= '2016-03-31') or (pe.startdate >= '2016-04-01' and pe.enddate <= '2016-06-30') ) and
((dv.sourceid in (select organisationunitid from _orgunitstructure where idlevel2 = 264)) ) and dv.deleted is false
limit 100000`;

describe('Analytics.aggregate', () => {
    let aggregate;
    let request;
    let mockApi;
    let fixture;

    beforeEach(() => {
        mockApi = MockApi.getApi();
        MockApi.mockClear();
        aggregate = new AnalyticsAggregate();
    });

    it('should not be allowed to be called without new', () => {
        expect(() => AnalyticsAggregate()).toThrowErrorMatchingSnapshot();
    });

    it('should add the mockApi onto the AnalyticsAggregate instance', () => {
        expect(aggregate.api).toBe(mockApi);
    });

    it('should use the api object when it is passed', () => {
        const apiMockObject = {};

        aggregate = new AnalyticsAggregate(apiMockObject);

        expect(aggregate.api).toBe(apiMockObject);
    });

    describe('.getDataValueSet()', () => {
        beforeEach(() => {
            aggregate = new AnalyticsAggregate(new MockApi());

            request = new AnalyticsRequest();

            request
                .addDataDimension(['fbfJHSPpUQD.pq2XI5kz2BY', 'fbfJHSPpUQD.PT59n8BQbqM'])
                .addPeriodDimension('LAST_MONTH')
                .addOrgUnitDimension('ImspTQPwCqd');

            fixture = fixtures.get('/api/analytics/dataValueSet');

            mockApi.get.mockReturnValue(Promise.resolve(fixture));
        });

        it('should be a function', () => {
            expect(aggregate.getDataValueSet).toBeInstanceOf(Function);
        });

        it('should resolve a promise with data', () => aggregate.getDataValueSet(request)
            .then((data) => {
                expect(data.dataValues.length).toEqual(fixture.dataValues.length);
            }));
    });

    describe('.getDebugSql()', () => {
        beforeEach(() => {
            aggregate = new AnalyticsAggregate(new MockApi());

            request = new AnalyticsRequest();

            request
                .addDataDimension(['fbfJHSPpUQD', 'cYeuwXTCPkU'])
                .addPeriodFilter(['2016Q1', '2016Q2'])
                .addOrgUnitFilter('O6uvpzGd5pu');

            mockApi.get.mockReturnValue(Promise.resolve(debugSqlFixture));
        });

        it('should be a function', () => {
            expect(aggregate.getDebugSql).toBeInstanceOf(Function);
        });

        it('should resolve a promise with data', () => aggregate.getDebugSql(request)
            .then((data) => {
                expect(data).toEqual(debugSqlFixture);
            }));
    });

    describe('.getRawData', () => {
        beforeEach(() => {
            aggregate = new AnalyticsAggregate(new MockApi());

            request = new AnalyticsRequest();

            request
                .addDataDimension(['fbfJHSPpUQD', 'cYeuwXTCPkU', 'Jtf34kNZhzP'])
                .addDimension('J5jldMd8OHv')
                .addDimension('Bpx0589u8y0')
                .addOrgUnitDimension(['O6uvpzGd5pu', 'fdc6uOvgoji'])
                .withStartDate('2016-01-01')
                .withEndDate('2016-01-31');

            fixture = fixtures.get('/api/analytics/rawData');

            mockApi.get.mockReturnValue(Promise.resolve(fixture));
        });

        it('should be a function', () => {
            expect(aggregate.getRawData).toBeInstanceOf(Function);
        });

        it('should resolve a promise with data', () => aggregate.getRawData(request)
            .then((data) => {
                expect(data.metaData.items).toEqual(fixture.metaData.items);
                expect(data.metaData.dimensions).toEqual(fixture.metaData.dimensions);
                expect(data.width).toEqual(0);
                expect(data.height).toEqual(0);
            }));
    });
});
