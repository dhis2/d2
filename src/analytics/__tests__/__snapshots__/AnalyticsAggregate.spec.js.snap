// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analytics.aggregate should not be allowed to be called without new 1`] = `"Class constructor AnalyticsAggregate cannot be invoked without 'new'"`;
